### ISSUE TITLE
InputQueue priority improvements

### ISSUE THREAD

**Issue by Noordfrees** [widelands/widelands#4374]
https://www.widelands.org/forum/topic/4994/?page=2

- Instead of three priority buttons, have a slider. This will make for a better click target and may be clearer regarding the meaning of the priority setting.
- Introduce additional states "Very Low" and "Highest" priority. The priorities are weighting constants defined in building.h and used in request.cc – the current ones are Low=2 Normal=4 High=8, we could additionally have VeryLow=0 (if possible, otherwise 1) and Highest=1024

**Comment by Noordfrees** [comment]
We should additionally consider converting the InputQueueDisplay to a subclass of UI::Box. It contains so many child elements with hardcoded positions and spacings that it is getting harder and harder to maintain. Only after #4376 is merged though.

**Comment by Noordfrees** [comment]
I am rewriting the InputQueueDisplay from scratch, so if we want a layout redesign now is the chance.

- We could have a conservative layout:
![q_classic](https://user-images.githubusercontent.com/48293446/97044818-f5410200-1574-11eb-9ced-f188199c857b.png)

- Or conservative with aligned priority sliders:
![q_align](https://user-images.githubusercontent.com/48293446/97044663-b317c080-1574-11eb-8fea-40ce26f47431.png)

- Or a completely redesigned look:
![q_modern](https://user-images.githubusercontent.com/48293446/97043873-71d2e100-1573-11eb-888d-b1ca5843725f.png)

- Optionally with aligned sliders:
![q_modern_align](https://user-images.githubusercontent.com/48293446/97044188-02a9bc80-1574-11eb-9448-802b54e4f40a.png)

My vote is for "redesigned with aligned sliders".
@widelands/developers

**Comment by stonerl** [comment]
My vote is for "redesigned with aligned sliders". And if possible aligned buttons and centered wares.

**Comment by Noordfrees** [comment]
> And if possible aligned buttons and centered wares.

Good idea, this would look either like this:
![q_modern_align_center1](https://user-images.githubusercontent.com/48293446/97046010-03901d80-1577-11eb-89af-1859f3a920b8.png)
Or alternatively like this:
![q_modern_align_center2](https://user-images.githubusercontent.com/48293446/97046020-068b0e00-1577-11eb-8d09-5d3cfed5c53b.png)

**Comment by Niektory** [comment]
These are rather big. It might be better to use something dropdown-like or spinner-like for them (with a conservative layout), or put them in a new tab.

**Comment by frankystone** [comment]
From an ui perspective it is necessary to let all wares have the same space in between. It indicates how many wares of each type is needed. If they are centered (space is added in between) this is not visible on first view.

**Comment by frankystone** [comment]
Just a side note: It would be nice to have such images with english translation.

**Comment by Noordfrees** [comment]
We can also make the slider a bit smaller and flush the wares left while still aligning the buttons:
![q_new](https://user-images.githubusercontent.com/48293446/97052413-41923f00-1581-11eb-824e-8cbf74a4cfee.png)

A dropdown or spinner would have the disadvantage of requiring (at least) one click extra. Another option might be to have a row of five radiobuttons instead (like what we have now, except that they'd all have the same size as the fill buttons).

> Just a side note: It would be nice to have such images with english translation.

Jau, wär dat nich düll. Is man för dat Upwiesen vun de Bruker-Schnittstee goor nich nödig. What's the need if the UI elements in question contain no text ;)
The images above all refer to a barbarian battle arena, except for this one which shows a barbarian trainingscamp. For the rating of the various designs, this implies …

---

Branch: https://github.com/Noordfrees/widelands/tree/4374-rewrite-inputqueuedisplay

**Comment by matthiakl** [comment]
> We can also make the slider a bit smaller and flush the wares left while still aligning the buttons:

I like this layout.

> From an ui perspective it is necessary to let all wares have the same space in between.

I agree. Additionally they should be aligned to one side for an easier check, how many wares of each type are present.

**Comment by Niektory** [comment]
There is not enough space for the vertical layout, as the training camp screenshot illustrates. Not being able to see all wares would be very annoying.

**Comment by Noordfrees** [comment]
Indeed it is annoying not to be able to see a window completely…
![shot0007](https://user-images.githubusercontent.com/48293446/97055056-43aacc80-1586-11eb-824b-8814c59f1f9e.png)

For now I have deliberately chosen a pretty small maximum size just to test that the scrollbox feature works properly. This is only a problem at very low resolutions.

**Comment by frankystone** [comment]
I would vote for 'conservative with aligned priority sliders'. 

![consercative](https://user-images.githubusercontent.com/48293446/97044663-b317c080-1574-11eb-8fea-40ce26f47431.png)

Since there are currently 5 Values to slide for (0,  4,  8, 16, 2048) the slider can be smaller imho: 5 * width of sliderbutton.

Instead of the background image, which adds too much colors imho, apply a color to the slider button? If it is left it will be reddish, if it is right it will be greenish? In normal state the slider button will be colored as the other buttons then.

**Comment by Noordfrees** [comment]
Another possible design would be not to use a slider after all:
![shot0009](https://user-images.githubusercontent.com/48293446/97077581-59031380-15e5-11eb-8fd9-5b58d2f8526e.png)
("conservative with aligned buttons and no slider")
(the icons are only placeholders)

**Comment by tppq** [comment]
> Introduce additional states "Very Low" and "Highest" priority.

I like this idea. The very low could be something that grabs stuff only if nobody else is asking for it. If implementing these using zero and max_int does it with modest amount of recoding, all the better.

**Comment by frankystone** [comment]
This also a nice idea, but i think a slider will fit better. While looking at 5 buttons on each row the brain has to calculate if it is pressed or not. A slider makes it much easier to the brain imho. And no extra images are needed. Just make the slider as small like the 5 buttons in this image :-)

Oh, and thanks for the localization :-)

**Comment by Noordfrees** [comment]
> > Introduce additional states "Very Low" and "Highest" priority.
> 
> I like this idea. The very low could be something that grabs stuff only if nobody else is asking for it. If implementing these using zero and max_int does it with modest amount of recoding, all the better.

With the current request implementation these priorities are only implemented as what they say. But if you can answer some of my questions at https://github.com/orgs/widelands/teams/developers/discussions/2 I may be able to implement this :)

> This also a nice idea, but i think a slider will fit better. While looking at 5 buttons on each row the brain has to calculate if it is pressed or not. A slider makes it much easier to the brain imho. And no extra images are needed. Just make the slider as small like the 5 buttons in this image :-)
> 
> Oh, and thanks for the localization :-)

"Conservative with small and aligned priority sliders"
![shot0010](https://user-images.githubusercontent.com/48293446/97092751-ca26e300-1646-11eb-8ae4-ff785a865c70.png)
(I remembered to prefix the command with `LANG=en` again just for you ;) )

**Comment by Niektory** [comment]
These work fine if you want to open the windows to fiddle with the settings and then close them. But I often leave windows open to keep an eye on the wares inside. I'm then only interested in the wares and everything else just gets in the way so I'd like it to take minimal space. It seems you have different preferences, so perhaps it should be toggleable somehow.

**Comment by frankystone** [comment]
Good point. Move them into a new tab "Wares priority settings"?

**Comment by Noordfrees** [comment]
A tab would make them too inaccessible for my liking, but a button to toggle the interface between fully expanded and show-only mode would be very good:
![shot0011](https://user-images.githubusercontent.com/48293446/97104911-ad76c380-16b7-11eb-8306-e0d37fcf3865.png)

**Comment by frankystone** [comment]
[I never want to read that widelands has already a complex UI anymore](https://www.widelands.org/forum/post/35251/) ;-)

**Comment by Noordfrees** [comment]
You call the non-collapsed state "not complex"!?
Is this statement meant as +1 or -1?

**Comment by Niektory** [comment]
> A tab would make them too inaccessible for my liking, but a button to toggle the interface between fully expanded and show-only mode would be very good:

This would work for me.

**Comment by matthiakl** [comment]
Then in show-only mode a small indicator should also show the current priority. Maybe like the current selector, but not clickable, so all priority states can be distinguished.

**Comment by Noordfrees** [comment]
![grafik](https://user-images.githubusercontent.com/48293446/97110254-6fd66280-16d8-11eb-815c-92ac6dbc7e3d.png)

**Comment by frankystone** [comment]
> You call the non-collapsed state "not complex"!?

Sorry if my statement wasn't clear. I just wanted to say that you are mostly arguing against some new buttons or other UI elements, whereas you are the one who makes the UI more complex, at least here. Please don't feel angry about my sentences. I really appreciate your work nevertheless :-)

> Is this statement meant as +1 or -1?

-1

But most other people like it, so i am outvoted.

**Comment by hessenfarmer** [comment]
Well on the one hand I understand the point of @frankystone. This is surely making the handling of the UI more complex which is always an issue that should be caefully evaluated.    
the reason why I am in favor of this solution is that it is better usable in 4k displays I guess as the slider is bigger and by this better manageable.   
To make the indicator less confusing we might have a blue button for the top to make it distinguishable. furthermore the colors should be visible on the slider as well to get the correlation  with the small indicators. 
we could even make the indicators a little wider and use them for unhiding the slider

**Comment by Noordfrees** [comment]
What exactly is it that makes the UI more complex? The replacement of one UI element (three buttons) with a different UI element (slider) is not additional complexity. The only new element here is the Collapse button, which in my opinion is not particularly obtrusive.

>  To make the indicator less confusing we might have a blue button for the top to make it distinguishable. furthermore the colors should be visible on the slider as well to get the correlation with the small indicators.

Like this?
![grafik](https://user-images.githubusercontent.com/48293446/97111911-bed4c580-16e1-11eb-9df1-3da9d864a97c.png)
(the slider looks a bit too noisy to me now)

>  we could even make the indicators a little wider and use them for unhiding the slider

-1 it is not logical in my opinion that this is revealed by clicking on a priority indicator rather than on the same button that was used to hide them

**Comment by hessenfarmer** [comment]
I think this is understandable now. But I agree the slider being noisy, so what about having only vertical lines nt he colors so it looks like a scala?

**Comment by Noordfrees** [comment]
How about this?
![image](https://user-images.githubusercontent.com/48293446/97118071-56e5a580-1708-11eb-9a3c-ad9ebbf6df1c.png)

Alternative design (which I like much less than the one above):
![image](https://user-images.githubusercontent.com/48293446/97117961-7cbe7a80-1707-11eb-933f-739c949e73bd.png)

**Comment by hessenfarmer** [comment]
the one above looks good to me.

**Comment by gunchleoc** [comment]
I think the ware capacity buttons should remain visible in the collapsed view. I use these a lot, while I use the priority rather seldom.

+1 for a horizontal, aligned design.

**Comment by Niektory** [comment]
> I think the ware capacity buttons should remain visible in the collapsed view. I use these a lot, while I use the priority rather seldom.

What if you could change the capacity by clicking on the ware icons? I think that would be quite convenient.

**Comment by Noordfrees** [comment]
> What if you could change the capacity by clicking on the ware icons? I think that would be quite convenient.

+1
We should use a double-click for that though, I am used to click-anywhere-and-drag to move windows.

**Comment by Niektory** [comment]
Double-clicking is double the effort, and inconsistent with how the rest of the UI works, so few people would try it.
I think there's still plenty of space to grab above, below and on the sides.
